### PR TITLE
feat: Add compatible on wasm

### PR DESCRIPTION
**Pull request by driftluo**
review #271 first

In order for wasm to run normally, all existing algorithms need to be able to compile to wasm. Therefore, replace the existing algorithm with the implemented by pure Rust and add corresponding tests

### CODE REVIEW

File: secio/src/crypto/wasm_compat.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    /// ```plain
+    /// +----------------------------------------+-----------------------+
+    /// | ENCRYPTED TEXT (length = output.len()) | TAG                   |
+    /// +----------------------------------------+-----------------------+
+    /// ```
+    pub fn decrypt(&mut self, input: &[u8]) -> Result<Vec<u8>, SecioError> {
+        nonce_advance(self.iv.as_mut());
+        self.cipher
+            .decrypt(Nonce::from_slice(self.iv.as_ref()), input)
+            .map_err(|_| SecioError::RingCryptoError)
```
Comment (quake) [inline secio/src/crypto/wasm_compat.rs:69]: suggest to renaming enum `SecioError::RingCryptoError` to `SecioError::CryptoError`

File: secio/src/codec/hmac_compat/mod.rs [hunk 2]
```diff
@@ -0,0 +1,45 @@
+#[cfg(not(target_os = "unknown"))]
```
Comment (quake) [inline secio/src/codec/hmac_compat/mod.rs:1]: should be `target_arch` ?
```suggestion
#[cfg(not(target_arch = "wasm32"))]
```

**Review by quake — CHANGES_REQUESTED** [review]

File: secio/src/crypto/wasm_compat.rs (hunk 1 above)
Comment (driftluo) [reply secio/src/crypto/wasm_compat.rs:69]: I don’t want to change the type for the time being, you can add todo here

File: secio/src/codec/hmac_compat/mod.rs (hunk 2 above)
Comment (driftluo) [reply secio/src/codec/hmac_compat/mod.rs:1]:
```
x86_64-pc-windows-gnu
x86_64-pc-windows-msvc
wasm32-unknown-emscripten
wasm32-unknown-unknown
wasm32-wasi
x86_64-unknown-linux-gnu
```

Obviously the third one is os, I think there is no problem writing this way

**Review by quake — APPROVED** [review]

File: secio/src/secp256k1_compat/mod.rs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+pub use wasm_compat::*;
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use rand::Rng;
+
+    #[test]
+    fn test_pk_to_sk() {
+        let mut message = [0; 64];
```
Comment (TheWaWaR) [inline secio/src/secp256k1_compat/mod.rs:19]: The variable name use `data` would be more accurate.

File: secio/src/secp256k1_compat/mod.rs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+        let signature = sign(&msg, &pk_native);
+
+        assert!(verify(&msg, &signature, &sk_native));
+
+        let msg_wasm = wasm_compat::message_from_slice(raw_msg.as_ref()).unwrap();
+
+        let signature_wasm = wasm_compat::signature_from_der(&signature_to_vec(signature)).unwrap();
+
+        assert!(wasm_compat::verify(&msg_wasm, &signature_wasm, &sk_wasm));
```
Comment (TheWaWaR) [inline secio/src/secp256k1_compat/mod.rs:42]: Should compare `sign(m, k) == wasm_compat::sign(m, k)` ?

**Review by TheWaWaR — APPROVED** [review]

**Pull request merged**